lhskim.py

```diff
@@ -0,0 +1,836 @@
+                ElectronPP(1,
+                    count_funcs=count_funcs),
+                LepHad(
+                    count_funcs=count_funcs),
+            ])
+
+            # set the event filters
+            self.filters['event'] = event_filters
+
+        # peek at first tree to determine which branches to exclude
+        with root_open(self.files[0]) as test_file:
+            test_tree = test_file.Get(self.metadata.treename)
+            ignore_branches = test_tree.glob(
+                hhbranches.REMOVE,
+                exclude=hhbranches.KEEP)
+            ignore_branches_output = test_tree.glob(
+                hhbranches.REMOVE_OUTPUT,
+                exclude=hhbranches.KEEP_OUTPUT)
+
+        # initialize the TreeChain of all input files
+        chain = TreeChain(
+            self.metadata.treename,
+            files=self.files,
+            ignore_branches=ignore_branches,
+            events=self.events,
+            onfilechange=onfilechange,
+            filters=event_filters,
+            cache=True,
+            cache_size=50000000,
+            learn_entries=100)
+
+        if local:
+            copied = [
+                'EventNumber',
+            ]
+
+            hh_buffer = TreeBuffer()
+            buffer = TreeBuffer()
+            for name, value in chain._buffer.items():
+                if name.startswith('hh_'):
+                    hh_buffer[name[3:]] = value
+                elif name in copied:
+                    buffer[name] = value
+            outtree.set_buffer(
+                hh_buffer,
+                create_branches=False,
+                visible=True)
+            outtree.set_buffer(
+                buffer,
+                create_branches=True,
+                visible=False)
+
+        else:
+            # additional decorations on existing objects
+            if year > 2011 and datatype in (datasets.MC, datasets.MCEMBED):
+                class Decorations(TreeModel):
+                    jet_ispileup = stl.vector('bool')
+
+                chain.set_buffer(Decorations(), create_branches=True)
+
+            # include the branches in the input chain in the output tree
+            # set branches to be removed in ignore_branches
+            outtree.set_buffer(
+                chain._buffer,
+                ignore_branches=ignore_branches + ignore_branches_output,
+                create_branches=True,
+                ignore_duplicates=True,
+                transfer_objects=True,
+                visible=False)
+
+        # define tree objects
+        define_objects(chain, year)
+
+        # create the MMC
+        mmc = mass.MMC(year=year)
+
+        # report which packages have been loaded
+        externaltools.report()
+
+        self.output.cd()
+
+        # The main event loop
+        # the event filters above are automatically run for each event and only
+        # the surviving events are looped on
+        for event in chain:
```

Comment:
Filters first. When the cutflow looks approximately right then we can start adding variables to the output tree.